### ISSUE TITLE
Allow access to StreamProviders from BootstrapProvider

### ISSUE THREAD

**Issue by osjimenez** [dotnet/orleans#587]
Exposing stream providers from within BootstrapProvider should not be a problem. 

Just need to ensure the correct initialization order and inject stream provider manager into Bootstrap Provider .

**Comment by ReubenBond** [comment]
:+1: I was just hoping to use this. Is it much work, @gabikliot?

**Comment by gabikliot** [comment]
I just checked and the initialization order is already OK. We do in this order:
1) load and init storage providers
2) load and init Stream Providers
3) start the silo membership. The silo is considered up and active after that.
4) Start Stream Providers - they may start pulling from the durable queues
5) load and init Bootstrap Providers

Therefore, passing stream providers to Bootstrap Providers when the latter start is safe.

Regarding how to inject:
@osjimenez wrote on Gitter:
> For now, I understand that the best way to pass StreamProviderManager to BootstrapProvider instances is add new parameter in 'Init' method of IProvider interface as:
`Task Init(string name, IProviderRuntime providerRuntime, IProviderConfiguration config, IStreamProviderManager streamProviderManager);`
Or, as this interface is more generic than `IBootstrapProvider`, I can add new method to `IBootstrapProvider` interface as:
`Task InitStreamProviderManager(IStreamProviderManager streamProviderManager);`

Indeed, I would not change the `IProvider.Init` method. Instead, I would add a new method on `IBootstrapProvider` interface:  `Task SetStreamProviderManager(IStreamProviderManager streamProviderManager);`
We will need to also add this method to BootstrapProviderManager which will call `pluginManager.GetProviders()` and call `SetStreamProviderManager` on each.
Should be pretty trivial.

**Comment by gabikliot** [comment]
@ReubenBond , I think we can nail it pretty quickly, between @osjimenez and myself.

**Comment by veikkoeeva** [comment]
@osjimenez, @gabikliot I wonder if were also OK to expose ``RuntimeIdentity`` in ``IProviderRuntime``? It looks like in debugger the concrete type of ``IProviderRuntime`` is (currently) ``SiloProviderRuntime`` and it has a property ``ExecutingSiloAddress`` which is in the same format. Maybe it could be renamed to ``RuntimeIdentity`` and exposed in the interface? This would help if one wants to do per silo initialization procedures, which is currently possible by calling a per-silo placed grain which returns the silo ID too, but it's a bit of a hack.

(I could check in the Orleans code, but I'm exhausted at the moment, just wanted to make a comment.)

**Comment by gabikliot** [comment]
@veikkoeeva , I think I agree, with the only caveat that `RuntimeIdentity`  is an opaque string and we would like to avoid exposing `SiloAddress` (`ExecutingSiloAddress`  is `SiloAddress`) to application code (to avoid any potential location dependent code). So I support exposing         `public string RuntimeIdentity` in the  IProviderRuntime   which will be identical to `Grain.RuntimeIdentity` (`SiloAddress.ToLongString()`).

**Comment by osjimenez** [comment]
I'm trying implement this with last Gabriel instructions. I was changed four files, you can view an explanation of this changes in this GIST:
https://gist.github.com/osjimenez/cbef04bec9bedd0f49c8

Unfortunatelly, this code isn't working now :( i receive a InvalidOperationException in my BootstrapProvider with this error message: ' Attempting to GetCurrentActivationData when not in an activation scope'

This is my provider code:
```csharp
    public class DemoBootstrapper : IBootstrapProvider
    {
        private IStreamProviderManager _streamProviderManager;
        public Task Init(string name, IProviderRuntime providerRuntime, IProviderConfiguration config)
        {
            Name = name;
            var pro = _streamProviderManager.GetProvider("demo") as IStreamProvider;
            var str = pro.GetStream<IEvent>(Guid.Empty, "demo");
            Factory.Create<EventPipe>().Subscribe(async evt =>
            {
                await str.OnNextAsync(evt);
            });
            return TaskDone.Done;
        }

        public string Name { get; private set; }
        public Task SetStreamProviderManager(IStreamProviderManager streamProviderManager)
        {
            this._streamProviderManager = streamProviderManager;
            return TaskDone.Done;
        }
    }
```
The InvalidOperationException throw in 'var str = pro.GetStream<IEvent>(Guid.Empty, "oka");' line.

What is wrong?

**Comment by gabikliot** [comment]
@osjimenez, I will look into it.

**Comment by osjimenez** [comment]
@gabikliot, you are in right !! i understand now that it makes no sense try to use a stream provider in initialization logic, i must use it after that, is obvious !

To avoid more misunderstands, i explain quickly you that i trying to do.

We need create a decoupled system to orchestration of multiple devices (fingerprint readers, card readers, plate recognition, payment systems, etc) These devices, more often now, are distributed by a large scale (across countries distributed some times). To accommodate these new requisites, we are creating a DDD+CQRS+ES solution, this is, based on commands and events patterns. 
I wish a solution that, with a single domain logic i could host in Azure, on-premise (client-server) and single process application, because each deployment has different size and has their  peculiarities.Orleans framework can give me now that i need, i think.

As you mentioned before, i may need a singleton grain in each silo to do that i want. To explain this properly i created a simple diagram:
![Diagram](http://i.snag.gy/0i46F.jpg)
Notes:
* EventPipe static class is an implementation of DomainEvents pattern (http://gorodinski.com/blog/2012/04/14/services-in-domain-driven-design-ddd/)
* For any silo instance, i need a singleton grain to subscribe to EventPipe and replay events to Orleans Stream Producer. This grains are (in example) Grain1, Grain3 and Grain4
* Silo2 not participate in the example sequence, is here only to show that EVERY silo must create a singleton class to replay domain events because if this grain is not created in one silo, the domain events can not propagate across silo instances, obviously

Example sequence:
* 1 - ViewModel in the UI sent a command to the Silo Cluster using Grain2 proxy (interface)
* 2 - Grain2 receives the command and process it in the domain model. This process raises some domain events that by EventPipe static class arrive to Grain1, then, this events will be replayed to all silos in the cluster using an Stream Producer
* 3 - Grain5 receives the event because is an Stream Consumer and refreshes the read model appropriately. Then, notify its observers using Orleans Client Observer feature
* 4 - In the UI, the Grain5 proxy raise an event that refreshes the UI

At this point, Do you think that i need create singleton grains? so i think yes. How i can define singleton grains and where i must initialize? This singleton grains will work as i described in the above sample?

I absolutely opened to any comment, idea, improvement, criticism, etc.

I hope not disturb you too much with my huge sample.
Thanks you very much !!!!

**Comment by gabikliot** [comment]
@osjimenez, I would love to understand your scenario and help you.
But I am having troubles understanding it: I don't see the diagram - the link seems to be dead. Also, what exactly is the Event Pipe?
In general, in Orleans, you should not be thinking in terms of static per silo objects and "reaching all silos". You should be thinking in terms of grains, which make no locality assumptions about their location.

> I wish a solution that, with a single domain logic i could host in Azure, on-premise (client-server) and single process application, because each deployment has different size and has their peculiarities.Orleans framework can give me now that i need, i think.

Orleans can definitely give you that.

**Comment by osjimenez** [comment]
I have repaired the above diagram link.

The EventPipe class is a static class to raise and subscribe to domain events. Is my attempt to implement the DomainEvents pattern: some links:
* Udi Dahan [Part one](http://www.udidahan.com/2008/02/29/how-to-create-fully-encapsulated-domain-models/), [Part two](http://www.udidahan.com/2008/08/25/domain-events-take-2/) and [Part three](http://www.udidahan.com/2009/06/14/domain-events-salvation/)
* [Martin Fowler](http://martinfowler.com/eaaDev/DomainEvent.html) (more clean to understand, i think) My EventPipe class is the same as Martin's EventProcessor class.

The main idea is create a event hub (in domain terms) to receive and manage all events in the whole system. I see my domain as a black box that accept commands as entries and return events as results. These results must be available to be processed by other domain actors and other services outside the domain.

I thought in only one grain per silo to replay the events, but I can try different approach removing the staticity of EventPipe class and instanciate one EventPipe by grain. I modify the diagram, now is much more simple:

![Diagram](http://img.ctrlv.in/img/15/07/10/559f808ab99a4.png)

Now, UI ViewModel sends a command to Grain1 and this command will be processed by Domain model logic, as a result, the Domain uses EventPipe to raise events. The grain is subscribed to its own EventPipe and replays events to Orleans Cluster using Orleans Stream feature. Then, Grain3 receives the events, refreshes the read model and notifies the UI to be refreshed.

I made EventPipe class static because in Udi Dahan (Part three) and some other internet code samples was so.

Is this approach better than previous diagram? Thinking about this new diagram, i have a doubt. Grain3 must be run since system start because is responsible to refresh the read model and must subscribe to every event that the domain model produces. If i create Grain3 on startup and subscribe to an Orleans Stream, Does runtime instantiate the grain when any other grain produce an event or i must do anything to keep this Grain3 alive?

Is interesting to me define well and thinking in deep about these concepts before starting develop because i don't know Orleans yet.

I hope i explained better now. A greeting.

**Comment by ReubenBond** [comment]
My reason for wanting this is that I want to do things outside of the Virtual Actor model occasionally. For example, I may want to invalidate a process-wide cache on each silo.

Does that sound like a valid scenario?

**Comment by gabikliot** [comment]
@osjimenez , thanks for the new explanation. It understand it much better now.
I think the 2nd approach, with event model per grain (and not one global static per silo) is more suitable for Orleans. The whole point about actors is actor isolation, not to share state between actors. There are sometimes exception to that rule, but I don't think the exception applies in your case.

Now to the question regarding grain 3 and how it subscribes to the stream.
First let me make sure that you do NOT mean that there will be one grain subscribing to all other grains. This would not scale well. There would be multiple "grains 3" in your scenario, reach subscribing to a different stream into which different sets of grain1-2 are publishing. Is that correct?>

Grain 3 can subscribe in a number of ways:
1) explicit subscription. For that oyu will need to send it some special grain command and tell it to subscribe.
2) implicit subscription, where the grain primary key will be the same as stream Guid. http://dotnet.github.io/orleans/Orleans-Streams/Streams-Programming-APIs "Explicit and Implicit Subscriptions".

Another option is to use stream in the client. Instead of using Observers to connect grain 3 back to the client you can skip grain 3 all together and let the client subscribe to this stream and produce the same way from grain 1 and 2. Streams works the same way on the client and in grains.

>If i create Grain3 on startup and subscribe to an Orleans Stream, Does runtime instantiate/activate the grain when any other grain produce an event or i must do anything to keep this Grain3 alive

The runtime instantiates/activates the grain when any other grain produce an event  and you do not need to do anything to keep this Grain3 alive

**Comment by gabikliot** [comment]
@ReubenBond , using Bootstrap provider for "a process-wide cache on each silo" totally makes sense and legit.
The question I have is do you still need the ability to access stream providers from the bootstrap provider? After the above discussion with @osjimenez , it seems like he does not really need it, since there is not much one can do with streams inside the bootstrap provider. 
Do you have a scenario where you need it? If yes, we can accept this PR based on @osjimenez's commit.

**Comment by ReubenBond** [comment]
@gabikliot awesome, thanks :)

I want to access stream providers so that I can use a stream to signal cache invalidation. It's a hack for development purposes, but one of our devs really doesn't like waiting for cache timeouts, so I wanted to give them a way to flush all caches for testing purposes.

**Comment by jthelin** [comment]
> The scenarios include any time we want to either break out of location transparency ensure something is executed on every silo.

I have also needed this type of "per-silo" grain type in some recent projects, and there are problems with any of the ways that can currently be achieved in user-level code.

Allowing a bootstrap provider to register a new System Target would be ideal for this, but those are only internal visibility so not re-usable outside the core runtime code.

Do you think it maybe a good time to dust off and expose the GraphPartitionPlacement strategy that was originally used in Project Horton?
http://richorama.github.io/blog/2015/01/02/grain-placement-in-orleans/ 
http://research.microsoft.com/en-us/projects/ldg/

**Comment by gabikliot** [comment]
OK, lets break it into smaller manageable items:
1) I think we should allow access to stream providers within bootstrap provider and allow to produce to the stream. Producing to the stream from the bootstrap provider  should be allowed just like making grain call from the bootstrap provider is. 
The first part is already implemented by @osjimenez and I will pick it up. The 2nd part will not work as of now I suspect, due to how we connect stream producers to grains (we install grain extension and it will not work as is for Bootstrap provider). I will look at this 2nd part. I hope it's easy to fix.

2) More general question about "per-silo" local grains. It has to do with much bigger and harder question about new type of grains, which will partially violate the virtual grain abstraction. This was also previously discussed in https://github.com/dotnet/orleans/issues/446 with @rore. 
In general, I am supportive of the idea of "per-silo" local grains or something similar in that spirit. The biggest issue I have with that is how to provide this new abstraction that will live in piece with the existing virtual actor abstraction. I suggest we take this question as a separate issue and come up with concrete suggestions, if you have any. I will say it again: The hard part in my opinion is to come up with the right abstraction. 

Do you guys agree?

**Comment by ReubenBond** [comment]
I agree, a clean way to implement per-silo grains is an important concern,
so we should do it right.

**Comment by jthelin** [comment]
Yes, agreed on both points.

**Comment by jthelin** [comment]
I also think we should try to keep PerSilo grain ideas separate from "replicated grains" discussed in #446 though. They are two mostly orthogonal features.

**Comment by gabikliot** [comment]
@ReubenBond and @osjimenez , I submitted #608 to allow producing to the stream from within Bootstrap provider.

**Comment by osjimenez** [comment]
Good work Gabriel, thanks.

I'm some days away of office . When back, i continue with the project implementing the last comments you made me.

A greeting.

**Comment by sergeybykov** [comment]
Sorry about joining the party late.

I'm concerned we are potentially mixing a few conceptually unrelated things here and may end up with an ugly picture if we rush a quick solution to plug a hole.

If we give a bootstrap provider access to streams, it suddenly becomes an implicitly addressable entity in the system. Even to become a producer, we would install under the covers a remotely addressable extension to receive notifications about new subscribers. If the bootstrap provider becomes a consumer, it will also be addressable, and potentially forever due to the virtual nature of Orleans streams.

If we step back, bootstrap providers were originally conceived as silo "autoexec.bat" kind of "scripts" to execute a set of custom initialization steps upon silo startup. I am concerned about the potential feature creep that will turn them into a new in-between abstraction open to abuse.

I understand the desire to let bootstrap providers post to a stream an "I'm up now" event or something, but I think the same can be easily be achieved via a call to a local grain acting as a proxy. Allowing bootstrap providers to subscribe to streams I think would be wrong for the reasons I mentioned above.

I'm convinced there are other, conceptually cleaner, means to achieve the desired functionality. Maybe it's exposing system targets to app layer. Maybe it's adding a broadcast feature to allow sending a call to all activated grains of a particular type. Maybe it's adding a fixed placement policy. Although the last one scares me a bit because we had it 3 (?) years ago for an internal customer, and ended up removing it in the end.

**Comment by gabikliot** [comment]
I disagree with @sergeybykov  point.
I don't think that "If we give a bootstrap provider access to streams, it suddenly becomes an implicitly addressable". It is not addressable by the public APIs, since it can only publish and not subscribe. 
I am not concerned about potential feature creep, since I don't think allowing to publish (as opposite to subscribe) is a new fundamental capability for  bootstrap providers.
Publish on a stream is like making a grain call, and I see no reason not to allow it.
Subscribe is different - it allows long term processing and this is indeed not what Bootstrap provider was meant to be in the first place.

Indeed, there will be ways to work around it, with an extra grain, but why adding another restriction/complications, for something that is not necessary and could be simplified.

I see the argument of not complicating the internal runtime, but after implementing it and needing to deal with the extra complexity, I actually don't think its a big complication. In theory any extra line is an added complexity, but in this case my call is that it is not much extra.

**Comment by jason-bragg** [comment]
Can do this with local client in 2.1.+
